odoo/tools/profiler.py

```diff
@@ -1,197 +1,401 @@
+        if event == 'line':
+            return
+        entry = {'event': event, 'frame': _format_frame(_frame)}
+        if event == 'call' and _frame.f_back:
+            # we need the parent frame to determine the line number of the call
+            entry['parent_frame'] = _format_frame(_frame.f_back)
+        self.add(entry, frame=_frame)
+        return self.hook
+
+    def _get_stack_trace(self, frame=None):
+        # Getting the full stack trace is slow, and not useful in this case.
+        # SyncCollector only saves the top frame and event at each call and
+        # recomputes the complete stack at the end.
+        return None
+
+    def post_process(self):
+        # Transform the evented traces to full stack traces. This processing
+        # could be avoided since speedscope will transform that back to
+        # evented anyway, but it is actually simpler to integrate into the
+        # current speedscope logic, especially when mixed with SQLCollector.
+        # We could improve it by saving as evented and manage it later.
+        stack = []
+        for entry in self._entries:
+            frame = entry.pop('frame')
+            event = entry.pop('event')
+            if event == 'call':
+                if stack:
+                    stack[-1] = entry.pop('parent_frame')
+                stack.append(frame)
+            elif event == 'return':
+                stack.pop()
+            entry['stack'] = stack[:]
+        super().post_process()
+
+
+class ExecutionContext:
+    """
+    Add some context on thread at current call stack level.
+    This context stored by collector beside stack and is used by Speedscope
+    to add a level to the stack with this information.
+    """
+    def __init__(self, **context):
+        self.context = context
+        self.stack_trace_level = None
+
+    def __enter__(self):
+        current_thread = threading.current_thread()
+        self.stack_trace_level = stack_size()
+        if not hasattr(current_thread, 'exec_context'):
+            current_thread.exec_context = {}
+        current_thread.exec_context[self.stack_trace_level] = self.context
+
+    def __exit__(self, *_args):
+        threading.current_thread().exec_context.pop(self.stack_trace_level)
 
-                n = k+1
-                while k+1 <= l and v['calls'][k+1]['callno'] != call['callno']:
-                    n += 1
-                if n >= l:
+
+class Profiler:
+    """
+    Context manager to use to start the recording of some execution.
+    Will save sql and async stack trace by default.
+    """
+    def __init__(self, db=..., path=None, collectors=None, profile_session=None, 
+                 description=None, disable_gc=False, params=None):
+        """
+        :param db: database name to use to save results.
+            Will try to define database automatically by default.
+            Use value ``None`` to not save results in a database.
+        :param path (str): path to use to save result
```

Comment:
type goes before parameter name, unparenthesized: https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists

Reply:
I actually got inspired from other odoo docstrings.
 
 ```python
 """
        :param str path : path to use to save result
        :param list<str|Collector>: collectors list of string and Collector object Ex: ['s...
        :param str profile_session: session description to use to reproup multiple p...
        :param str description: description of the current profiler Suggestion: (rou...
        :param bool disable_gc: flag to disable gc durring profiling (usefull to avoid gc ...
        :param dict params: parameters usable by collectors (like frame interval...
"""
```

I usually agree to respect conventions but in this case it is less readable after because parameters name are unaligned. What do you think?

Reply:
> What do you think?

That it's wrong no matter how much you prefer it.

Info fields are parsed by tooling, they're parsed by sphinx, and they're parsed by editors (pycharm at the very least), and the format is specifically `:param [type] name:`.

If you don't like it then either:

* don't use type annotations at all
* move the type annotations to PEP484-style type hints (in the formal parameters list):

      def __init__(self, db=..., path: str = None, collectors=None, profile_session=None, 

  that works fine for PyCharm (and I assume Sphinx though I didn't check)

* use the `:type:` directive to split name and type:

    :param path: ...
    :param collectors: ...
    :param profile_session: ...
    :type path: str
    :type collectors: list
    :type profile_session: list